### ISSUE TITLE
Add constructors (ReadOnly)Span<T> (ref T)

### ISSUE THREAD

**Issue by atykhyy** [dotnet/runtime#39425]
## Background and Motivation

Currently, it is impossible to construct a `(ReadOnly)Span<T>` from a `ref T` without unsafe code. In some circumstances, enabling compilation of unsafe code on an assembly may trigger additional security reviews or be otherwise undesirable. The existing constructors from `void*, int` and `ref T, int` are of course dangerous because the count argument is not checked, the former constructor being marked as unsafe and the latter being internal. However, constructing a span of _one_ element from `ref T` is a simple and safe operation, and there seems to be no reason not to allow it. 

## Proposed API

```diff
namespace System
{
    public readonly ref struct Span<T>
    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Span (ref T ptr) : this (ref ptr, 1) {}
    }
    public readonly ref struct ReadOnlySpan<T>
    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ReadOnlySpan (ref T ptr) : this (ref ptr, 1) {}
    }
}
```

## Usage Examples

Adding these two constructors would help bridge the gap between `ref` and Span-based APIs, e.g. when writing serialization code, and avoid allocations/unsafe code, for example
```csharp
class MyWrapperStream : Stream
{
    // implementation is writing to APIs that expect spans
    public override void WriteByte (byte value)
    {
        // no allocations or unsafe code needed
        Write (new ReadOnlySpan<byte> (ref value)) ;
    }
}
```

## Alternative Designs

The current workaround is to use unsafe code and the dangerous `void*, int` constructor.

## Risks

Since `(ReadOnly)Span<T>` are ref structs, and don't allow overruns, there is no risk of memory corruption.

**Comment by SingleAccretion** [comment]
What would this API do in case of an "invalid" `ref`? How would it detect it?
````C#
var emptyArray = new T[0];
var span = new Span<T>(ref emptyArray[0]);
````

**Comment by stephentoub** [comment]
This already exists as MemoryMarshal.CreateSpan.

**Comment by atykhyy** [comment]
So it does, albeit not in netstandard2.0 (but that ship has long sailed). `MemoryMarshal.CreateSpan` however is unsafe, because the `count` parameter is not checked, and is accordingly stuck in `System.Runtime.InteropServices` together with other methods that can wreck the process despite not being `unsafe`. The proposed constructors would be safe.

**Comment by jkotas** [comment]
> The proposed constructors would be safe.

It would not be safe. The rules that C# compiler enforces for ref structs would not kick in for this correctly. For example:
`Span<byte> f() { byte b; return new Span<byte>(ref b); }` would compile fine, but it is clearly unsafe.

This can be fixed by adding more rules or special cases to the rules that C# compiler enforces for ref structs, but that is a lot more work than just adding an trivial API.

**Comment by atykhyy** [comment]
Hm. If that's the case, then I agree, but the compiler already realizes that this variant of your example leaks a local reference:
```
Span<byte> Foo ()
{
    Span<byte> foo = stackalloc byte[16] ;
    return foo ; // error CS8352
}
```

**Comment by IllidanS4** [comment]
It always seemed to me that the compiler should know that when a ref struct is constructed from a reference, the reference *could* be captured in there somehow.

**Comment by jkotas** [comment]
https://github.com/dotnet/csharplang/issues/1147#issuecomment-534657035 and related posts have discussion about what it would take.

**Comment by atykhyy** [comment]
I've looked at the code in Roslyn binder where it checks escape levels and I agree that special-casing some ref struct object creation expressions would not be trivial, albeit incomparably simpler than the general problem of ref fields in ref structs. However, since `Span`s are already subject to escape analysis and are allowed as members of ref structs, perhaps such special-casing would help with dotnet/csharplang#1147 as well? Let existing escape analysis code do the heavy lifting. The users can declare intent - i.e. whether they intend to capture the reference or merely use it, as a counter or something - unambiguously by passing a ref vs a `Span` into a ref struct's constructor, and use `Span` (or perhaps a new span-like struct wrapping just a pointer, to avoid the overhead of storing a count that equals 1) instead of ref fields. A `Span` has a well-defined default value, so no additional complications around initialization would seem to be involved. The jitter would optimize away the struct methods and the output code should be as good as with straight pointers.

**Comment by jaredpar** [comment]
@jkotas is correct here. The ref safety rules in C# specifically call out that they rely on this constructor not existing 

https://github.com/dotnet/csharplang/blob/master/proposals/csharp-7.2/span-safety.md#span-constructor

As a part of implementing ref fields though it's very likely that the restriction would be removed. Some other discussion on that is available here

https://github.com/dotnet/csharplang/blob/master/proposals/csharp-7.2/span-safety.md#length-one-spant-over-ref-values

**Comment by N0D4N** [comment]
I believe it is a duplicate of #67445 and should be closed since #67445 is already implemented
